go/stats/gauges.go

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2025 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package stats
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// Gauges tracks gauge values over time by sampling them at regular intervals.
+// Unlike Rates which tracks operation rates, Gauges tracks actual values.
+type Gauges struct {
+	mu         sync.Mutex
+	timeStamps *RingInt64
+	history    map[string]*RingInt64    // Historical sampled values
+	current    map[string]*atomic.Int64 // Current gauge values
+	samples    int
+	interval   time.Duration
+	ctx        context.Context
+	cancel     context.CancelFunc
+}
+
+// NewGauges creates a new Gauges object that samples gauge values at regular intervals.
+// samples: number of historical samples to keep
+// interval: how often to sample the current gauge values
+func NewGauges(samples int, interval time.Duration) *Gauges {
+	if interval < 1*time.Second && interval != -1*time.Second {
+		panic("interval too small")
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	g := &Gauges{
+		timeStamps: NewRingInt64(samples + 1),
+		history:    make(map[string]*RingInt64),
+		current:    make(map[string]*atomic.Int64),
+		samples:    samples + 1,
+		interval:   interval,
+		ctx:        ctx,
+		cancel:     cancel,
+	}
+
+	// Start background sampling goroutine
+	if interval > 0 {
+		go g.track()
+	}
+
+	return g
+}
+
+// Set updates the current value for the named gauge.
+// This value will be sampled at the next interval.
+func (g *Gauges) Set(name string, value int64) {
+	g.mu.Lock()
+	gauge, ok := g.current[name]
+	if !ok {
+		gauge = &atomic.Int64{}
+		g.current[name] = gauge
+	}
+	g.mu.Unlock()
+
+	gauge.Store(value)
+}
+
+// Stop terminates the background sampling goroutine.
+func (g *Gauges) Stop() {
+	g.cancel()
+}
```

Comment:
nit: should we wait here for `track()` to return? using a channel?

Reply:
I don't think so. track() will return when the context is cancelled here:
```
case <-g.ctx.Done():
    return
```